File: lib/src/core/release_descriptor.dart
import "dart:convert";

/// Parsed `release.json` metadata for one platform-specific update artifact.
///
/// Descriptors use schema version 3 and point at one platform-specific artifact
/// plus the install strategy needed by the native helper. Optional delta
/// artifact metadata can be published ahead of runtime support; the updater
/// continues to choose the full artifact until delta verification and patch
/// application are implemented.
class ReleaseDescriptor {
  /// Creates release metadata for one downloadable artifact.
  const ReleaseDescriptor({
    required this.schemaVersion,
    required this.packageId,
    required this.appName,
    required this.version,
    required this.buildNumber,
    required this.platform,
    required this.channel,
    required this.artifact,
    required this.install,
    required this.minimumUpdaterVersion,
    required this.generatedAt,
    this.minimumOS = const {},
    this.deltaArtifacts = const [],
    this.signature,
  });

  /// Parses and validates a schema-v3 release descriptor from JSON.
  factory ReleaseDescriptor.fromJson(Map<String, dynamic> json) {
    final descriptor = ReleaseDescriptor(
      schemaVersion: json["schemaVersion"] as int? ?? 0,
      packageId: json["packageId"] as String? ?? "",
      appName: json["appName"] as String? ?? "",
      version: json["version"] as String? ?? "",
      buildNumber: json["buildNumber"] as int?,
      platform: json["platform"] as String? ?? "",
      channel: json["channel"] as String? ?? "stable",
      artifact: ReleaseArtifact.fromJson(
        json["artifact"] as Map<String, dynamic>? ?? const {},
      ),
      install: ReleaseInstall.fromJson(
        json["install"] as Map<String, dynamic>? ?? const {},
      ),
      minimumOS: _parseMinimumOS(json["minimumOS"]),
      deltaArtifacts: _parseDeltaArtifacts(json["deltaArtifacts"]),
      signature: json["signature"] == null
          ? null
          : ReleaseSignature.fromJson(
              json["signature"] as Map<String, dynamic>,
            ),
      minimumUpdaterVersion: json["minimumUpdaterVersion"] as String? ?? "",
      generatedAt: DateTime.parse(
        json["generatedAt"] as String? ?? "1970-01-01T00:00:00Z",
      ),
    )..validate();
    return descriptor;
  }

  /// Descriptor schema version. This package currently supports version 3.
  final int schemaVersion;

  /// Stable package identifier used to match releases to an app.
  final String packageId;

  /// Human-readable app name shown by update UI and release tooling.
  final String appName;

  /// Semantic app version for this release.
  final String version;

  /// Optional platform build number used as a same-version tiebreaker.
  final int? buildNumber;

  /// Target platform identifier, such as `macos`, `windows`, or `linux`.
  final String platform;

  /// Release channel, such as `stable`.
  final String channel;

  /// Downloadable artifact described by this release.
  final ReleaseArtifact artifact;

  /// Install strategy for the staged artifact.
  final ReleaseInstall install;

  /// Optional detached descriptor signature metadata.
  final ReleaseSignature? signature;

  /// Minimum updater package version expected to handle this descriptor.
  final String minimumUpdaterVersion;

  /// Optional minimum OS versions keyed by platform.
  final Map<String, String> minimumOS;

  /// Optional descriptor metadata for future delta update artifacts.
  final List<ReleaseDeltaArtifact> deltaArtifacts;

  /// UTC timestamp for descriptor generation.
  final DateTime generatedAt;

  /// Converts this descriptor to the schema-v3 JSON shape.
  Map<String, dynamic> toJson() {
    return {
      "schemaVersion": schemaVersion,
      "packageId": packageId,
      "appName": appName,
      "version": version,
      if (buildNumber != null) "buildNumber": buildNumber,
      "platform": platform,
      "channel": channel,
      "artifact": artifact.toJson(),
      "install": install.toJson(),
      if (signature != null) "signature": signature!.toJson(),
      "minimumUpdaterVersion": minimumUpdaterVersion,
      if (minimumOS.isNotEmpty) "minimumOS": minimumOS,
      if (deltaArtifacts.isNotEmpty)
        "deltaArtifacts": [
          for (final artifact in deltaArtifacts) artifact.toJson(),
        ],
      "generatedAt": generatedAt.toUtc().toIso8601String(),
    };
  }

  /// Returns the minimum OS value for [platform], when descriptor metadata has
  /// one.
  String? minimumOSForPlatform(String platform) {
    return minimumOS[platform];
  }

  /// Returns the stable JSON payload used for signing and verification.
  ///
  /// When [signature] is present, its value is blanked before sorting so the
  /// signature signs the descriptor envelope rather than itself.
  Map<String, dynamic> toCanonicalSignatureJson() {
    final json = toJson();
    final existingSignature = signature;
    if (existingSignature != null) {
      json["signature"] = existingSignature.copyWith(value: "").toJson();
    }
    return sortJsonValue(json) as Map<String, dynamic>;
  }

  /// Encodes [toCanonicalSignatureJson] as UTF-8 bytes.
  List<int> canonicalSignatureBytes() {
    return utf8.encode(jsonEncode(toCanonicalSignatureJson()));
  }

  /// Validates required schema, identity, artifact, and install fields.
  void validate() {
    if (schemaVersion != 3) {
      throw FormatException(
        "Unsupported release descriptor schema version: $schemaVersion",
      );
    }
    if (packageId.trim().isEmpty) {
      throw const FormatException("release.json packageId is required.");
    }
    if (version.trim().isEmpty) {
      throw const FormatException("release.json version is required.");
    }
    if (platform.trim().isEmpty) {
      throw const FormatException("release.json platform is required.");
    }
    if (buildNumber != null && buildNumber! < 0) {
      throw const FormatException(
        "release.json buildNumber must be zero or greater when provided.",
      );
    }
    artifact.validate();
    for (final deltaArtifact in deltaArtifacts) {
      deltaArtifact.validate();
    }
    install.validate(platform: platform, artifactKind: artifact.kind);
  }
}

Map<String, String> _parseMinimumOS(Object? value) {
  if (value == null) {
    return const {};
  }
  if (value is! Map) {
    throw const FormatException("release.json minimumOS must be an object.");
  }

  final minimumOS = <String, String>{};
  for (final entry in value.entries) {
    final platform = entry.key.toString().trim();
    final version = entry.value?.toString().trim() ?? "";
    if (platform.isEmpty || version.isEmpty) {
      throw const FormatException(
        "release.json minimumOS entries must use non-empty strings.",
      );
    }
    minimumOS[platform] = version;
  }
  return Map.unmodifiable(minimumOS);
}

List<ReleaseDeltaArtifact> _parseDeltaArtifacts(Object? value) {
  if (value == null) {
    return const [];
  }
  if (value is! List) {
    throw const FormatException("release.json deltaArtifacts must be a list.");
  }

  final artifacts = <ReleaseDeltaArtifact>[];
  for (final entry in value) {
    if (entry is! Map) {
      throw const FormatException(
        "release.json deltaArtifacts entries must be objects.",
      );
    }
    artifacts.add(
      ReleaseDeltaArtifact.fromJson(Map<String, dynamic>.from(entry)),
    );
  }
  return List.unmodifiable(artifacts);
}

List<String> _parseStringList(Object? value, String displayName) {
  if (value == null) {
    return const [];
  }
  if (value is! List) {
    throw FormatException("release.json $displayName must be a list.");
  }
  return List.unmodifiable([for (final entry in value) entry.toString()]);
}

/// Download metadata for the artifact referenced by a descriptor.
class ReleaseArtifact {
  /// Creates artifact metadata.
  const ReleaseArtifact({
    required this.kind,
    required this.url,
    required this.sha256,
    required this.length,
  });

  /// Parses artifact metadata from the descriptor `artifact` object.
  factory ReleaseArtifact.fromJson(Map<String, dynamic> json) {
    return ReleaseArtifact(
      kind: json["kind"] as String? ?? "",
      url: Uri.parse(json["url"] as String? ?? ""),
      sha256: json["sha256"] as String? ?? "",
      length: json["length"] as int? ?? -1,
    );
  }

  /// Artifact type. Version 3 descriptors support `zip`, `dmg`, and
  /// `pkgInstaller`.
  final String kind;

  /// Absolute URL for downloading the artifact.
  final Uri url;

  /// Expected lowercase hexadecimal SHA-256 digest.
  final String sha256;

  /// Expected artifact length in bytes.
  final int length;

  /// Converts this artifact to descriptor JSON.
  Map<String, dynamic> toJson() {
    return {
      "kind": kind,
      "url": url.toString(),
      "sha256": sha256,
      "length": length,
    };
  }

  /// Validates artifact kind, digest shape, and byte length.
  void validate() {
    if (kind != "zip" && kind != "dmg" && kind != "pkgInstaller") {
      throw FormatException("Unsupported release artifact kind: $kind");
    }
    if (!RegExp(r"^[0-9a-f]{64}$").hasMatch(sha256)) {
      throw const FormatException(
        "release.json artifact.sha256 must be 64 lowercase hex characters.",
      );
    }
    if (length < 0) {
      throw const FormatException("release.json artifact.length is required.");
    }
  }
}

/// Descriptor metadata for a future delta update artifact.
class ReleaseDeltaArtifact {
  /// Creates delta artifact metadata.
  const ReleaseDeltaArtifact({
    required this.fromVersion,
    required this.kind,
    required this.url,
    required this.sha256,
    required this.length,
  });

  /// Parses delta artifact metadata from a descriptor entry.
  factory ReleaseDeltaArtifact.fromJson(Map<String, dynamic> json) {
    return ReleaseDeltaArtifact(
      fromVersion: json["fromVersion"] as String? ?? "",
      kind: json["kind"] as String? ?? "",
      url: Uri.parse(json["url"] as String? ?? ""),
      sha256: json["sha256"] as String? ?? "",
      length: json["length"] as int? ?? -1,
    );
  }

  /// Source app version that this delta can patch from.
  final String fromVersion;

  /// Delta artifact type. Version 3 descriptors currently reserve `bsdiff`.
  final String kind;

  /// Absolute URL for downloading the delta artifact.
  final Uri url;

  /// Expected lowercase hexadecimal SHA-256 digest.
  final String sha256;

  /// Expected delta artifact length in bytes.
  final int length;

  /// Converts this delta artifact to descriptor JSON.
  Map<String, dynamic> toJson() {
    return {
      "fromVersion": fromVersion,
      "kind": kind,
      "url": url.toString(),
      "sha256": sha256,
      "length": length,
    };
  }

  /// Throws because runtime delta verification is not implemented yet.
  void ensureRuntimeSupported() {
    throw UnsupportedError(
      "Delta update artifacts are not supported yet; use the full zip "
      "artifact.",
    );
  }

  /// Validates delta metadata shape without enabling runtime application.
  void validate() {
    if (fromVersion.trim().isEmpty) {
      throw const FormatException(
        "release.json deltaArtifacts.fromVersion is required.",
      );
    }
    if (kind != "bsdiff") {
      throw FormatException("Unsupported release delta artifact kind: $kind");
    }
    if (!RegExp(r"^[0-9a-f]{64}$").hasMatch(sha256)) {
      throw const FormatException(
        "release.json deltaArtifacts.sha256 must be 64 lowercase hex "
        "characters.",
      );
    }
    if (length < 0) {
      throw const FormatException(
        "release.json deltaArtifacts.length is required.",
      );
    }
  }
}

/// Install metadata for a staged release artifact.
class ReleaseInstall {
  /// Creates install metadata with the requested [strategy].
  const ReleaseInstall({required this.strategy, this.macosDmg, this.macosPkg});

  /// Parses install metadata from the descriptor `install` object.
  factory ReleaseInstall.fromJson(Map<String, dynamic> json) {
    return ReleaseInstall(
      strategy: json["strategy"] as String? ?? "",
      macosDmg: json["macosDmg"] == null
          ? null
          : ReleaseMacOSDmgInstall.fromJson(
              json["macosDmg"] as Map<String, dynamic>,
            ),
      macosPkg: json["macosPkg"] == null
          ? null
          : ReleaseMacOSPkgInstall.fromJson(
              json["macosPkg"] as Map<String, dynamic>,
            ),
    );
  }

  /// Native helper strategy used to install the staged artifact.
  final String strategy;

  /// macOS DMG update metadata.
  final ReleaseMacOSDmgInstall? macosDmg;

  /// macOS PKG installer metadata.
  final ReleaseMacOSPkgInstall? macosPkg;

  /// Converts this install metadata to descriptor JSON.
  Map<String, dynamic> toJson() {
    return {
      "strategy": strategy,
      if (macosDmg != null) "macosDmg": macosDmg!.toJson(),
      if (macosPkg != null) "macosPkg": macosPkg!.toJson(),
    };
  }

  /// Validates that an install strategy is present.
  void validate({String? platform, String? artifactKind}) {
    if (strategy.trim().isEmpty) {
      throw const FormatException("release.json install.strategy is required.");
    }
    if (artifactKind == "dmg") {
      if (platform != "macos" || strategy != "wholeBundleReplace") {
        throw const FormatException(
          "release.json dmg artifacts are only supported for macos "
          "wholeBundleReplace.",
        );
      }
      if (macosDmg == null) {
        throw const FormatException(
          "release.json install.macosDmg is required for dmg artifacts.",
        );
      }
      macosDmg!.validate();
    }
    if (artifactKind == "pkgInstaller") {
      if (platform != "macos" || strategy != "pkgInstaller") {
        throw const FormatException(
          "release.json pkgInstaller artifacts require install.strategy "
          "pkgInstaller on macos.",
        );
      }
      if (macosPkg == null) {
        throw const FormatException(
          "release.json install.macosPkg is required for pkgInstaller.",
        );
      }
      macosPkg!.validate();
    }
    if (strategy == "pkgInstaller" && artifactKind != "pkgInstaller") {
      throw const FormatException(
        "release.json pkgInstaller strategy is only supported for "
        "pkgInstaller artifacts.",
      );
    }
  }
}

/// macOS DMG update metadata from `release.json`.
class ReleaseMacOSDmgInstall {
  /// Creates DMG update metadata.
  const ReleaseMacOSDmgInstall({
    required this.appBundleName,
    required this.verifyPrimarySignature,
  });

  /// Parses DMG update metadata.
  factory ReleaseMacOSDmgInstall.fromJson(Map<String, dynamic> json) {
    return ReleaseMacOSDmgInstall(
      appBundleName: json["appBundleName"] as String? ?? "",
      verifyPrimarySignature: json["verifyPrimarySignature"] as bool? ?? true,
    );
  }

  /// Simple `.app` bundle name expected in the mounted DMG.
  final String appBundleName;

  /// Whether to assess the DMG primary signature before mounting.
  final bool verifyPrimarySignature;

  /// Converts this policy to descriptor JSON.
  Map<String, dynamic> toJson() {
    return {
      "appBundleName": appBundleName,
      "verifyPrimarySignature": verifyPrimarySignature,
    };
  }

  /// Validates the DMG policy.
  void validate() {
    if (!appBundleName.endsWith(".app") || appBundleName.contains("/")) {
      throw const FormatException(
        "release.json install.macosDmg.appBundleName must be a simple .app "
        "name.",
      );
    }
  }
}

/// macOS PKG installer metadata from `release.json`.
class ReleaseMacOSPkgInstall {
  /// Creates PKG installer metadata.
  const ReleaseMacOSPkgInstall({
    required this.launchMode,
    required this.expectedPackageIds,
    required this.relaunchAfterInstall,
  });

  /// Parses PKG installer metadata.
  factory ReleaseMacOSPkgInstall.fromJson(Map<String, dynamic> json) {
    return ReleaseMacOSPkgInstall(
      launchMode: json["launchMode"] as String? ?? "installerApp",
      expectedPackageIds: _parseStringList(
        json["expectedPackageIds"],
        "install.macosPkg.expectedPackageIds",
      ),
      relaunchAfterInstall: json["relaunchAfterInstall"] as bool? ?? false,
    );
  }

  /// Native handoff mode. The runtime currently supports Installer.app only.
  final String launchMode;

  /// Package identifiers expected inside the installer package metadata.
  final List<String> expectedPackageIds;

  /// Whether the app should relaunch after Installer.app completes.
  final bool relaunchAfterInstall;

  /// Converts this policy to descriptor JSON.
  Map<String, dynamic> toJson() {
    return {
      "launchMode": launchMode,
      "expectedPackageIds": expectedPackageIds,
      "relaunchAfterInstall": relaunchAfterInstall,
    };
  }

  /// Validates the PKG policy.
  void validate() {
    if (launchMode != "installerApp") {
      throw const FormatException(
        "release.json install.macosPkg.launchMode must be installerApp.",
      );
    }
    if (expectedPackageIds.isEmpty) {
      throw const FormatException(
        "release.json install.macosPkg.expectedPackageIds must not be empty.",
      );
    }
  }
}

/// Signature metadata embedded in a release descriptor.
class ReleaseSignature {
  /// Creates descriptor signature metadata.
  const ReleaseSignature({
    required this.algorithm,
    required this.publicKeyId,
    required this.value,
  });

  /// Parses signature metadata from the descriptor `signature` object.
  factory ReleaseSignature.fromJson(Map<String, dynamic> json) {
    return ReleaseSignature(
      algorithm: json["algorithm"] as String? ?? "",
      publicKeyId: json["publicKeyId"] as String? ?? "",
      value: json["value"] as String? ?? "",
    );
  }

  /// Signature algorithm identifier, currently expected to be `ed25519`.
  final String algorithm;

  /// Identifier for the pinned public key that should verify [value].
  final String publicKeyId;

  /// Base64-encoded signature bytes.
  final String value;

  /// Converts this signature metadata to descriptor JSON.
  Map<String, dynamic> toJson() {
    return {"algorithm": algorithm, "publicKeyId": publicKeyId, "value": value};
  }

  /// Returns a copy with a replacement signature [value].
  ReleaseSignature copyWith({String? value}) {
    return ReleaseSignature(
      algorithm: algorithm,
      publicKeyId: publicKeyId,
      value: value ?? this.value,
    );
  }
}

/// Recursively sorts JSON map keys for deterministic descriptor signing.
Object? sortJsonValue(Object? value) {
  if (value is Map) {
    final sorted = <String, dynamic>{};
    final keys = value.keys.map((key) => key.toString()).toList()..sort();
    for (final key in keys) {
      sorted[key] = sortJsonValue(value[key]);
    }
    return sorted;
  }

  if (value is List) {
    return value.map(sortJsonValue).toList(growable: false);
  }

  return value;
}

File: macos/desktop_updater/Sources/DesktopUpdaterKit/Runtime/ReleaseDescriptor.swift
import Foundation

public struct ReleaseArtifact {
    public let kind: String
    public let url: URL
    public let sha256: String
    public let length: Int64
    public let rawJSON: [String: Any]

    init(json: [String: Any]) throws {
        kind = try runtimeString(json, "kind")
        url = try runtimeAbsoluteURL(json, "url")
        sha256 = try runtimeString(json, "sha256")
        guard sha256.range(
            of: "^[0-9a-f]{64}$",
            options: .regularExpression
        ) != nil else {
            throw RuntimeError.outcome(
                .invalidDescriptor,
                message: "Artifact SHA-256 must use lowercase hexadecimal."
            )
        }
        guard let parsedLength = try runtimeOptionalInt64(json["length"]),
              parsedLength >= 0
        else {
            throw RuntimeError.outcome(
                .invalidDescriptor,
                message: "Artifact length must not be negative."
            )
        }
        length = parsedLength
        rawJSON = [
            "kind": kind,
            "url": url.absoluteString,
            "sha256": sha256,
            "length": length,
        ]
    }
}

public struct ReleaseInstall {
    public let strategy: String
    public let normalizedJSON: [String: Any]
    public let rawJSON: [String: Any]

    init(json: [String: Any], platform: String, artifactKind: String) throws {
        strategy = try runtimeTrimmedString(json, "strategy")
        guard ["zip", "dmg", "pkgInstaller", "innoInstaller"]
            .contains(artifactKind)
        else {
            throw RuntimeError.outcome(
                .unsupportedArtifactKind,
                message: "Unsupported artifact kind \(artifactKind)."
            )
        }

        var normalized: [String: Any] = ["strategy": strategy]
        var pkgWireLaunchMode: String?
        if let rawInno = json["inno"] {
            let metadata = try runtimeDictionary(rawInno)
            let arguments = try runtimeStringList(
                metadata["silentArgs"],
                key: "silentArgs"
            )
            let inheritDirectory = try runtimeOptionalBool(
                metadata["inheritInstallDirectory"],
                default: true
            )
            let logFileName = try runtimeOptionalString(
                metadata["logFileName"],
                default: "desktop_updater_inno_install.log"
            )
            let relaunch = try runtimeOptionalBool(
                metadata["relaunchAfterInstall"],
                default: true
            )
            let elevation = try runtimeOptionalString(
                metadata["requiresElevation"],
                default: "auto"
            )
            var authenticode: [String: Any] = ["required": false]
            if let rawAuthenticode = metadata["authenticode"] {
                let policy = try runtimeDictionary(rawAuthenticode)
                let required = try runtimeOptionalBool(
                    policy["required"],
                    default: false
                )
                let thumbprints = try runtimeStringList(
                    policy["sha256Thumbprints"],
                    key: "sha256Thumbprints",
                    default: []
                )
                authenticode = ["required": required]
                if !thumbprints.isEmpty {
                    authenticode["sha256Thumbprints"] = thumbprints
                }
            }
            normalized["inno"] = [
                "silentArgs": arguments,
                "inheritInstallDirectory": inheritDirectory,
                "logFileName": logFileName,
                "relaunchAfterInstall": relaunch,
                "requiresElevation": elevation,
                "authenticode": authenticode,
            ]
        }
        if let rawDMG = json["macosDmg"] {
            let metadata = try runtimeDictionary(rawDMG)
            normalized["macosDmg"] = [
                "appBundleName": try runtimeString(metadata, "appBundleName"),
                "verifyPrimarySignature": try runtimeOptionalBool(
                    metadata["verifyPrimarySignature"],
                    default: true
                ),
            ]
        }
        if let rawPKG = json["macosPkg"] {
            let metadata = try runtimeDictionary(rawPKG)
            let rawLaunchMode = try runtimeOptionalString(
                metadata["launchMode"],
                default: "installerApp"
            )
            pkgWireLaunchMode = rawLaunchMode
            normalized["macosPkg"] = [
                "launchMode": rawLaunchMode == "installerApp"
                    ? "privilegedInstallerTool"
                    : rawLaunchMode,
                "expectedPackageIds": try runtimeStringList(
                    metadata["expectedPackageIds"],
                    key: "expectedPackageIds",
                    default: []
                ),
                "relaunchAfterInstall": try runtimeOptionalBool(
                    metadata["relaunchAfterInstall"],
                    default: false
                ),
            ]
        }

        if strategy == "pkgInstaller", artifactKind != "pkgInstaller" {
            throw RuntimeError.outcome(
                .invalidDescriptor,
                message: "PKG strategy requires a PKG artifact."
            )
        }
        if strategy == "innoInstaller",
           (platform != "windows" || artifactKind != "innoInstaller")
        {
            throw RuntimeError.outcome(
                .invalidDescriptor,
                message: "Inno strategy requires a Windows installer artifact."
            )
        }
        if artifactKind == "dmg" {
            guard platform == "macos",
                  strategy == "wholeBundleReplace",
                  let metadata = normalized["macosDmg"] as? [String: Any],
                  let appName = metadata["appBundleName"] as? String,
                  appName.hasSuffix(".app"),
                  !appName.contains("/")
            else {
                throw RuntimeError.outcome(
                    .invalidDescriptor,
                    message: "Invalid macOS DMG install metadata."
                )
            }
        }
        if artifactKind == "pkgInstaller" {
            guard platform == "macos",
                  strategy == "pkgInstaller",
                  let metadata = normalized["macosPkg"] as? [String: Any],
                  metadata["launchMode"] as? String == "privilegedInstallerTool",
                  let packageIds = metadata["expectedPackageIds"] as? [String],
                  !packageIds.isEmpty
            else {
                throw RuntimeError.outcome(
                    .invalidDescriptor,
                    message: "Invalid macOS PKG install metadata."
                )
            }
        }
        if artifactKind == "innoInstaller" {
            let innoMetadata = normalized["inno"] as? [String: Any]
            let authenticode = innoMetadata?["authenticode"]
                as? [String: Any]
            let thumbprints =
                (authenticode?["sha256Thumbprints"] as? [String]) ?? []
            guard platform == "windows",
                  strategy == "innoInstaller",
                  let metadata = innoMetadata,
                  let arguments = metadata["silentArgs"] as? [String],
                  !arguments.isEmpty,
                  arguments.contains("/VERYSILENT") || arguments.contains("/SILENT"),
                  let logFileName = metadata["logFileName"] as? String,
                  !logFileName.trimmingCharacters(
                      in: .whitespacesAndNewlines
                  ).isEmpty,
                  !logFileName.contains("/"),
                  !logFileName.contains("\\"),
                  let elevation = metadata["requiresElevation"] as? String,
                  ["auto", "always", "never"].contains(elevation),
                  let authenticode,
                  let required = authenticode["required"] as? Bool,
                  !required || !thumbprints.isEmpty,
                  thumbprints.allSatisfy({
                      $0.range(
                          of: "^[0-9A-Fa-f]{64}$",
                          options: .regularExpression
                      ) != nil
                  })
            else {
                throw RuntimeError.outcome(
                    .invalidDescriptor,
                    message: "Invalid Windows Inno install metadata."
                )
            }
        }
        normalizedJSON = normalized
        if let pkgWireLaunchMode,
           var wirePKG = normalized["macosPkg"] as? [String: Any]
        {
            wirePKG["launchMode"] = pkgWireLaunchMode
            normalized["macosPkg"] = wirePKG
        }
        rawJSON = normalized
    }
}

public struct ReleaseSignature {
    public let algorithm: String
    public let publicKeyId: String
    public let value: String
}

public struct ReleaseDescriptor {
    public let schemaVersion: Int
    public let packageId: String
    public let appName: String
    public let version: String
    public let buildNumber: Int64?
    public let platform: String
    public let channel: String
    public let artifact: ReleaseArtifact
    public let install: ReleaseInstall
    public let signature: ReleaseSignature?
    public let minimumUpdaterVersion: String
    public let minimumOS: [String: String]
    public let deltaArtifacts: [[String: Any]]
    public let generatedAt: Date
    public let rawJSON: [String: Any]

    public init(jsonData: Data) throws {
        let json = try runtimeDictionary(
            JSONSerialization.jsonObject(with: jsonData)
        )
        guard try runtimeInt(json, "schemaVersion") == 3 else {
            throw RuntimeError.outcome(
                .invalidDescriptor,
                message: "Release descriptor must use schema version 3."
            )
        }
        schemaVersion = 3
        packageId = try runtimeTrimmedString(json, "packageId")
        appName = try runtimeString(json, "appName")
        version = try runtimeTrimmedString(json, "version")
        _ = try DesktopVersion(version)
        buildNumber = try runtimeOptionalInt64(json["buildNumber"])
        if let buildNumber, buildNumber < 0 {
            throw RuntimeError.outcome(
                .invalidDescriptor,
                message: "Build number must not be negative."
            )
        }
        platform = try runtimeTrimmedString(json, "platform")
        channel = try json["channel"].map { _ in
            try runtimeTrimmedString(json, "channel")
        } ?? "stable"
        artifact = try ReleaseArtifact(
            json: runtimeDictionary(json["artifact"] as Any)
        )
        install = try ReleaseInstall(
            json: runtimeDictionary(json["install"] as Any),
            platform: platform,
            artifactKind: artifact.kind
        )
        if artifact.kind == "pkgInstaller", buildNumber == nil {
            throw RuntimeError.outcome(
                .invalidDescriptor,
                message: "PKG release descriptors require a build number."
            )
        }
        if let signatureJSON = json["signature"] {
            let value = try runtimeDictionary(signatureJSON)
            signature = ReleaseSignature(
                algorithm: try runtimeString(value, "algorithm"),
                publicKeyId: try runtimeString(value, "publicKeyId"),
                value: try runtimeStringAllowingEmpty(value, "value")
            )
        } else {
            signature = nil
        }
        minimumUpdaterVersion = try runtimeString(
            json,
            "minimumUpdaterVersion"
        )
        _ = try DesktopVersion(minimumUpdaterVersion)
        if let minimumOSJSON = json["minimumOS"] {
            let values = try runtimeDictionary(minimumOSJSON)
            var parsed: [String: String] = [:]
            for (key, value) in values {
                let normalizedKey = key.trimmingCharacters(
                    in: .whitespacesAndNewlines
                )
                guard let rawVersion = value as? String else {
                    throw RuntimeError.outcome(
                        .invalidDescriptor,
                        message: "Invalid minimum OS policy."
                    )
                }
                let version = rawVersion.trimmingCharacters(
                    in: .whitespacesAndNewlines
                )
                guard !normalizedKey.isEmpty, !version.isEmpty
                else {
                    throw RuntimeError.outcome(
                        .invalidDescriptor,
                        message: "Invalid minimum OS policy."
                    )
                }
                parsed[normalizedKey] = version
            }
            minimumOS = parsed
        } else {
            minimumOS = [:]
        }
        if let deltaJSON = json["deltaArtifacts"] {
            guard let values = deltaJSON as? [Any] else {
                throw RuntimeError.outcome(
                    .invalidDescriptor,
                    message: "Delta artifacts must be a list."
                )
            }
            deltaArtifacts = try values.map {
                let delta = try runtimeDictionary($0)
                let fromVersion = try runtimeString(delta, "fromVersion")
                guard try runtimeString(delta, "kind") == "bsdiff",
                      let url = URL(
                          string: try runtimeString(delta, "url")
                      ),
                      url.scheme?.isEmpty == false,
                      try runtimeString(delta, "sha256").range(
                          of: "^[0-9a-f]{64}$",
                          options: .regularExpression
                      ) != nil,
                      let length = try runtimeOptionalInt64(delta["length"]),
                      length >= 0,
                      !fromVersion.isEmpty
                else {
                    throw RuntimeError.outcome(
                        .invalidDescriptor,
                        message: "Invalid delta artifact metadata."
                    )
                }
                return [
                    "fromVersion": fromVersion,
                    "kind": try runtimeString(delta, "kind"),
                    "url": url.absoluteString,
                    "sha256": try runtimeString(delta, "sha256"),
                    "length": length,
                ]
            }
        } else {
            deltaArtifacts = []
        }
        guard let generated = RuntimeISO8601.date(
            try runtimeString(json, "generatedAt")
        ) else {
            throw RuntimeError.outcome(
                .invalidDescriptor,
                message: "Invalid descriptor generation time."
            )
        }
        generatedAt = generated
        var normalized: [String: Any] = [
            "schemaVersion": schemaVersion,
            "packageId": packageId,
            "appName": appName,
            "version": version,
            "platform": platform,
            "channel": channel,
            "artifact": artifact.rawJSON,
            "install": install.rawJSON,
            "minimumUpdaterVersion": minimumUpdaterVersion,
            "generatedAt": RuntimeISO8601.string(generatedAt),
        ]
        if let buildNumber {
            normalized["buildNumber"] = buildNumber
        }
        if let signature {
            normalized["signature"] = [
                "algorithm": signature.algorithm,
                "publicKeyId": signature.publicKeyId,
                "value": signature.value,
            ]
        }
        if !minimumOS.isEmpty {
            normalized["minimumOS"] = minimumOS
        }
        if !deltaArtifacts.isEmpty {
            normalized["deltaArtifacts"] = deltaArtifacts
        }
        rawJSON = normalized
    }

    public func canonicalSignatureBytes() throws -> Data {
        var canonical: [String: Any] = [
            "schemaVersion": schemaVersion,
            "packageId": packageId,
            "appName": appName,
            "version": version,
            "platform": platform,
            "channel": channel,
            "artifact": artifact.rawJSON,
            "install": install.rawJSON,
            "minimumUpdaterVersion": minimumUpdaterVersion,
            "generatedAt": RuntimeISO8601.string(generatedAt),
        ]
        if let buildNumber {
            canonical["buildNumber"] = buildNumber
        }
        if let signature {
            canonical["signature"] = [
                "algorithm": signature.algorithm,
                "publicKeyId": signature.publicKeyId,
                "value": "",
            ]
        }
        if !minimumOS.isEmpty {
            canonical["minimumOS"] = minimumOS
        }
        if !deltaArtifacts.isEmpty {
            canonical["deltaArtifacts"] = deltaArtifacts
        }
        return try JSONSerialization.data(
            withJSONObject: canonical,
            options: [.sortedKeys, .withoutEscapingSlashes]
        )
    }

    public func bindingOutcome(
        indexItem: ReleaseIndexItem,
        expectedPackageId: String
    ) -> RuntimeOutcome? {
        if packageId != expectedPackageId {
            return .packageIdentityMismatch
        }
        if version != indexItem.version ||
            buildNumber != indexItem.buildNumber ||
            platform != indexItem.platform ||
            channel != indexItem.channel
        {
            return .invalidDescriptor
        }
        return nil
    }
}

private func runtimeOptionalBool(
    _ value: Any?,
    default defaultValue: Bool
) throws -> Bool {
    guard let value else { return defaultValue }
    guard let number = value as? NSNumber,
          CFGetTypeID(number) == CFBooleanGetTypeID()
    else {
        throw RuntimeError.outcome(
            .invalidDescriptor,
            message: "Expected boolean install metadata."
        )
    }
    return number.boolValue
}

private func runtimeOptionalString(
    _ value: Any?,
    default defaultValue: String
) throws -> String {
    guard let value else { return defaultValue }
    guard let string = value as? String else {
        throw RuntimeError.outcome(
            .invalidDescriptor,
            message: "Expected string install metadata."
        )
    }
    return string
}

private func runtimeStringList(
    _ value: Any?,
    key: String,
    default defaultValue: [String]? = nil
) throws -> [String] {
    guard let value else {
        if let defaultValue { return defaultValue }
        throw RuntimeError.outcome(
            .invalidDescriptor,
            message: "Expected list \(key)."
        )
    }
    guard let values = value as? [Any] else {
        throw RuntimeError.outcome(
            .invalidDescriptor,
            message: "Expected list \(key)."
        )
    }
    guard values.allSatisfy({ $0 is String }) else {
        throw RuntimeError.outcome(
            .invalidDescriptor,
            message: "Expected string entries in list \(key)."
        )
    }
    return values.compactMap { $0 as? String }
}

private func runtimeStringAllowingEmpty(
    _ source: [String: Any],
    _ key: String
) throws -> String {
    guard let value = source[key] as? String else {
        throw RuntimeError.outcome(
            .invalidDescriptor,
            message: "Expected string \(key)."
        )
    }
    return value
}
